### PR TITLE
[FLINK-15418][table-planner-blink] Set FlinkRelDistribution in StreamExecMatchRule

### PR DESCRIPTION
**Pull request by libenchao**
## What is the purpose of the change

This pr fix that StreamExecMatchRule not sets FlinkRelDistribution trait.

## Brief change log

1, Set FlinkRelDistribution in StreamExecMatchRule.

## Verifying this change

This change added tests and can be verified as follows:

MatchRecognizeITCase.testPartitionByWithParallelSource

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/MatchRecognizeITCase.scala [hunk 1]
```diff
@@ -416,6 +416,54 @@ class MatchRecognizeITCase(backend: StateBackendMode) extends StreamingWithState
...
+    val data = new mutable.MutableList[(String, Long, Int, Int)]
+    data.+=(("ACME", 1L, 19, 1))
+    data.+=(("ACME", 2L, 17, 2))
+    data.+=(("ACME", 3L, 13, 3))
+    data.+=(("ACME", 4L, 20, 4))
+    val parallelSource = new ParallelCollectionWrapper[(String, Long, Int, Int)](
+      data, 0, data.length)
+
+    val t = env.fromParallelCollection(parallelSource)
+      .assignAscendingTimestamps(tickerEvent => tickerEvent._2)
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/MatchRecognizeITCase.scala]: Apply `.setParallelism(4)` to reshuffle data to reproduce the problem, in this way, we don't need to introduce `ParallelCollectionWrapper`.

**Comment by libenchao** [comment]
@wuchong Thanks for the review, I've updated the code according to suggestions.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecMatchRule.scala [hunk 2]
```diff
@@ -57,16 +57,19 @@ class StreamExecMatchRule
   override def convert(rel: RelNode): RelNode = {
     val logicalMatch: FlinkLogicalMatch = rel.asInstanceOf[FlinkLogicalMatch]
     val traitSet: RelTraitSet = rel.getTraitSet.replace(FlinkConventions.STREAM_PHYSICAL)
-
     val partitionKeys = logicalMatch.getPartitionKeys
-    val newTraitSet = if (partitionKeys != null && !partitionKeys.isEmpty) {
-      traitSet.replace(FlinkRelDistribution.hash(logicalMatch.getPartitionKeys.toList))
+
+    val requiredDistribution = if (partitionKeys != null && !partitionKeys.isEmpty) {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecMatchRule.scala]: Nit: `partitionKeys` is never null.

Comment (libenchao) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecMatchRule.scala]: removed.

**Pull request closed without merging**